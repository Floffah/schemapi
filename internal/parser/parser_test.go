package parser

import (
	"errors"
	"github.com/gkampitakis/go-snaps/snaps"
	"os"
	"strings"
	"testing"
)

func TestParserSnapshot(t *testing.T) {
	files, err := os.ReadDir("../../tests")

	if err != nil {
		t.Fatalf("failed to read examples directory: %v", err)
	}

	for _, file := range files {
		if !strings.HasSuffix(file.Name(), ".sapi") {
			continue
		}

		t.Run(file.Name(), func(t *testing.T) {
			data, err := os.ReadFile("../../tests/" + file.Name())

			if err != nil {
				t.Fatalf("failed to read %s: %v", file.Name(), err)
			}

			parser := NewParser(string(data))
			rootNode, err := parser.Parse()

			if err != nil {
				// if error is of type ParseError, we can provide more context
				var parseErr *ParserError
				if errors.As(err, &parseErr) {
					t.Fatalf("\n%s:%d:%d: %s", file.Name(), parseErr.Line, parseErr.Col, parseErr.Message)
				} else {
					t.Fatalf("failed to parse %s: %v", file.Name(), err)
				}
			}

			snaps.MatchSnapshot(t, rootNode)
		})
	}
}
