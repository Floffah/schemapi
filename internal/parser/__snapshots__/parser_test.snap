[TestParserSnapshot/callable.sapi - 1]
&parser.RootNode{
    Children: {
        &parser.CallableNode{
            Type:       &parser.IdentifierNode{Name:"http.get"},
            Identifier: &parser.PathNode{
                parts: {"healthz"},
            },
            Params: {
            },
            Children: {
            },
        },
    },
}
---

[TestParserSnapshot/full.sapi - 1]
&parser.RootNode{
    Children: {
        &parser.CallableNode{
            Type:       &parser.IdentifierNode{Name:"http.get"},
            Identifier: &parser.PathNode{
                parts: {"healthz"},
            },
            Params: {
            },
            Children: {
                &parser.CallableDefinitionNode{
                    Type:   &parser.IdentifierNode{Name:"query"},
                    Params: {
                        &parser.IdentifierNode{Name:"string"},
                        &parser.StringNode{Value:"name"},
                    },
                    Children: &parser.DictionaryNode{
                        Entries: {
                            {
                                Identifier: &parser.IdentifierNode{Name:"description"},
                                Value:      &parser.StringNode{Value:"Name of the service to check health for"},
                            },
                            {
                                Identifier: &parser.IdentifierNode{Name:"default"},
                                Value:      &parser.StringNode{Value:"default-service"},
                            },
                        },
                    },
                },
                &parser.CallableDefinitionNode{
                    Type:   &parser.IdentifierNode{Name:"response"},
                    Params: {
                        &parser.StringNode{Value:"OK"},
                    },
                    Children: &parser.DictionaryNode{
                        Entries: {
                            {
                                Identifier: &parser.IdentifierNode{Name:"status"},
                                Value:      &parser.NumberNode{Value:"200"},
                            },
                            {
                                Identifier: &parser.IdentifierNode{Name:"body"},
                                Value:      &parser.DictionaryNode{
                                    Entries: {
                                        {
                                            Identifier: &parser.IdentifierNode{Name:"message"},
                                            Value:      &parser.StringNode{Value:"Service {name} is healthy"},
                                        },
                                    },
                                },
                            },
                        },
                    },
                },
                &parser.CallableDefinitionNode{
                    Type:   &parser.IdentifierNode{Name:"response"},
                    Params: {
                        &parser.StringNode{Value:"Service Not Found"},
                    },
                    Children: &parser.DictionaryNode{
                        Entries: {
                            {
                                Identifier: &parser.IdentifierNode{Name:"status"},
                                Value:      &parser.NumberNode{Value:"404"},
                            },
                            {
                                Identifier: &parser.IdentifierNode{Name:"body"},
                                Value:      &parser.DictionaryNode{
                                    Entries: {
                                        {
                                            Identifier: &parser.IdentifierNode{Name:"message"},
                                            Value:      &parser.StringNode{Value:"Service {name} not found"},
                                        },
                                    },
                                },
                            },
                        },
                    },
                },
                &parser.CallableDefinitionNode{
                    Type:   &parser.IdentifierNode{Name:"response"},
                    Params: {
                        &parser.StringNode{Value:"Internal Server Error"},
                    },
                    Children: &parser.DictionaryNode{
                        Entries: {
                            {
                                Identifier: &parser.IdentifierNode{Name:"status"},
                                Value:      &parser.NumberNode{Value:"500"},
                            },
                            {
                                Identifier: &parser.IdentifierNode{Name:"body"},
                                Value:      &parser.DictionaryNode{
                                    Entries: {
                                        {
                                            Identifier: &parser.IdentifierNode{Name:"message"},
                                            Value:      &parser.StringNode{Value:"An unexpected error occurred"},
                                        },
                                    },
                                },
                            },
                        },
                    },
                },
            },
        },
    },
}
---
